app/services/katello/pulp3/alternate_content_source.rb

```diff
@@ -0,0 +1,126 @@
+require "pulpcore_client"
+module Katello
+  module Pulp3
+    class AlternateContentSource
+      include Katello::Pulp3::ServiceCommon
+      attr_accessor :acs
+      attr_accessor :smart_proxy
+
+      def initialize(acs, smart_proxy)
+        @acs = acs
+        @smart_proxy = smart_proxy
+      end
+
+      def api
+        @api ||= ::Katello::Pulp3::Repository.api(smart_proxy, @acs.content_type)
+      end
+
+      def generate_backend_object_name
+        "#{acs.label}-#{smart_proxy.url}-#{rand(9999)}"
+      end
+
+      def smart_proxy_acs
+        ::Katello::SmartProxyAlternateContentSource.find_by(alternate_content_source_id: acs.id, smart_proxy_id: smart_proxy.id)
+      end
+
+      def remote_options
+        remote_options = {
+          tls_validation: acs.verify_ssl,
+          name: generate_backend_object_name,
+          url: acs.base_url,
+          policy: 'on_demand',
+          proxy_url: acs.http_proxy&.url,
+          proxy_username: acs.http_proxy&.username,
+          proxy_password: acs.http_proxy&.password,
+          total_timeout: Setting[:sync_connect_timeout]
+        }
+        if acs.content_type == ::Katello::Repository::FILE_TYPE && acs.subpaths.empty? && !remote_options[:url].end_with?('/PULP_MANIFEST')
+          remote_options[:url] = acs.base_url + '/PULP_MANIFEST'
+        end
+        if !acs.upstream_username.blank? && !acs.upstream_password.blank?
+          remote_options.merge!(username: acs.upstream_username,
+                               password: acs.upstream_password)
```

Comment:
I remember we had an issue with remote where users weren't able to clear username/password. Does this tackle that?

Reply:
Ah yeah this was added before that fix, I'll update it.